src/d3a/models/market/__init__.py

```diff
@@ -264,7 +264,8 @@ def get_bids_offers_trades(self):
 
 
 def validate_authentic_bid_offer_pair(bid, offer, clearing_rate, selected_energy):
-    assert bid.energy >= selected_energy
-    assert offer.energy >= selected_energy
-    assert (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= clearing_rate
-    assert (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= offer.energy_rate
+    if not (bid.energy >= selected_energy and
+            offer.energy >= selected_energy and
+            (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= clearing_rate and
+            (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= offer.energy_rate):
+        raise Exception
```

Comment:
This is a clear improvement over the previous assertions, that's great 👍 
I have another remark to this: it is highly encouraged to avoid raising/catching the base `Exception` class. It would be much better to create a specific exception for this and possibly catch it later on. It could be something generic like `MarketException` or something more specific like `BidOfferValidationException`, or anything really to avoid the base `Exception` :)

Reply:
Some more information: https://rules.sonarsource.com/python/RSPEC-112#:~:text=Raising%20instances%20of%20Exception%20and,prone%20and%20difficult%20to%20maintain.

Reply:
The goal of this exception-handling here is to make sure no transaction is posted if any other one will fail, a runtime exception of any type can be thrown here for some other reason besides validation, so we either keep it the way it is now or add new Exception type and add another broad exception handler in the calling code, what do you think ?

Reply:
> a runtime exception of any type can be thrown here for some other reason besides validation

What kind of exceptions do you think might occur here? In theory, the method does not need to deal with all possible types of exceptions, because other parts of the code should have already dealt with those. For example, you will in general expect `bid.energy_rate` to exist and never raise an `AttributeError`. If you really want to catch all possible types of exceptions (which is not ideal, but it's an option sometimes) you could do something like:
```python
try:
    if not (...):
except Exception as ex:
    raise SomeCustomException(ex) from ex

class SomeCustomException(Exception):
    """Exception raised when this and that happens."""
```
So you would still raise a custom exception, and the other parts of the code will know where it comes from. Also, it will contain the message raised by the original exception, which is necessary to properly debug errors.
At this point, you can either catch it using its name, or with a broad `Exception` (not ideal, but sometimes not easily avoidable).

Reply:
I suppose that, to avoid crashes on running simulations, there's a main entry point for d3a where all exceptions are dealt with, catching a broad `Exception`.